src/__tests__/useDebugLogger.spec.ts

```diff
@@ -0,0 +1,72 @@
+/**
+ * @jest-environment jsdom
+ */
+import { cleanup } from "@testing-library/react";
+import { renderHook } from "@testing-library/react-hooks";
+import { useState } from "react";
+import TestRenderer from "react-test-renderer";
+import { useDebugLogger } from "../hooks/useDebugLogger";
+
+const { act } = TestRenderer;
+
+const useHook = () => {
+  const [value, setValue] = useState(0);
+
+  useDebugLogger("info", value);
```

Comment:
> I really think we should reconsider JSON.stringify here and just let users figure out how exactly they want to debug log their data.

They can also do it + have all of the benefits from above :)

I am AFK this week but let me get back on this next week.

Reply:
Let's do a `useDebugLogger` and a `useDeepDebugLoggerExperimental` and we will use the `JSON.stringify` in the second one. We will make it non experimental if everyone likes it. What say? @capJavert

Reply:
@imbhargav5 I am sorry but I do not agree with the above. I strongly think that it should go with current implementations and I have put all of the benefits above of which all are valid I no con vas presented against them.

For your other notes above:
> It gets less performant the deeper the data is if its an object

It really does not, you would need to have really really large objects to catch any kind of performance impact. Also in production `JSON.streingify` is skipped  https://github.com/imbhargav5/rooks/pull/606/files#diff-73ac7d3c0660e42d9efa2ced58a7cde408c5d05e8cce841cada0303dfb5e195cR21 because `useDebugLogger` is also skipped in production.

For example I tried pulling 25mb JSON file https://raw.githubusercontent.com/json-iterator/test-data/master/large-file.json, parse it to JS and then `JSON.stringify` it. It took 70-80ms on my old 2015 i5 macbook pro. So I don't see how performance is the issue here.

> It is not useful if the data is a primitive

It is not but for primitives works exactly as if primitive is passed.

If you still do not agree I think I'll respectfully contribute this as my own package. Thanks for your comments.

Reply:
@capJavert Totally get where you are coming from. It's probably frustrating to have this conversation over a period of a few weeks for you. Apologies for that. Busy timings for both of us. 

I agree about perf. We don't really have to care about that since it's dev only. Thanks for correcting me on that. 
Regarding including JSON.stringify within the hook itself. I strongly disagree with that. It has to be opt-in. It can't be a part of the hook itself the simplest reason being, when I want to log a value, I don't expect a stringified value to begin with.

## Other reasons to not use JSON.stringify
1. It formats values to null when we don't need to 

![Screenshot 2021-11-23 at 1 33 49 AM](https://user-images.githubusercontent.com/2936644/142927416-7eeb32e7-ba6c-4496-a180-6964a5643fac.png)

2. It works differently with dates
![Screenshot 2021-11-23 at 1 35 59 AM](https://user-images.githubusercontent.com/2936644/142927670-3df21351-700c-4a9b-82ab-8fe14af6b247.png)

3. Doesn't work well with functions 
![Screenshot 2021-11-23 at 1 38 03 AM](https://user-images.githubusercontent.com/2936644/142927967-7a1959f1-fa7e-40f0-815d-30e9fa45b691.png)
4. Throws errors with objects with circular references. 
![Screenshot 2021-11-23 at 1 38 39 AM](https://user-images.githubusercontent.com/2936644/142928043-7ad6093b-d845-49e3-b4a7-7a889bd32aea.png)

![Screenshot 2021-11-23 at 1 43 21 AM](https://user-images.githubusercontent.com/2936644/142928663-db27424a-2513-498c-9779-89e13ebd3caa.png)

Feel free to disagree but I can't support adding JSON.stringify as default behaviour even if the above 4 points were not there. Mainly because logging and stringifying are two different things and the name `useDebugLogger` doesn't semantically refer to a stringified logger. 

### Summary

Bottom line is. We should not worry about all of this for a simple hook by trying to over deliver. By keeping the signature simple, we can allow users to use JSON.stringify if they need to easily but also just be able to log whatver they want with the most simplest function they already know. `console.log`.

`const useDebugLogger = (value, logLevel?, format? ) => {}` is how we can do it. This way we don't have the headache of opinionating a format option. And the user can decide how they want to log it. It can be as simple as

```
useDebugLogger({a: 1}, "info");
useDebugLogger({a:1}, JSON.stringify)

//import query string from 'some-package'
useDebugLogger({a:1},"info",(data)=>`data is ${data}`)
useDebugLogger({a:1},"info",querystring.stringify)
``` 

We can try a different signature or a different usage if you have more ideas, but JSON.stringify needs to be arg. Can't be default.

Reply:
Thanks for the response. I think you missed the part about JSON.stringify again, hook does not use it to display value but just as the internal resolver of equality, for example it does not matter if it changes date because again internally it will just resolve it as 2 same date strings.

In respects to equality Functions is a valid example of where it does not work but again benefits I think outweigh that.

I am not interested in complicating the syntax so I will have to take this few steps back and publish it as standalone package.

I am sorry because including it in rooks would give the hook more reach and usage. But well, sometimes it does not work out.

Thanks for the notes and keep up the good work with rooks. 💪

Reply:
@capJavert Ah. I forgot the purpose we were using JSON.stringify. And drew the wrong conclusion. Honestly, what have I been smoking in this thread. However, I spent some time on this today and there is a better solution than `JSON.stringify` IMO (unless I am still very stupid today).

Good luck with your hook! I am sure it will come out great. 

I am still going to implement a `useDebugLogger` and add you as a contributor as gratitude for your effort with this PR. The internals will be slightly different though. Thanks for your hardwork. Means a lot!